File: spkmeans.py
#imports:
import sys
import enum
import numpy as np
import pandas as pd
import spkmeans as sp
import matplotlib.pyplot as plt

#"goal" enum:
class goalEnum (enum.Enum):
    wam = "wam"
    ddg = "ddg"
    lnorm = "lnorm"
    jacobi = "jacobi"
    spk = "spk"

#methods:
def getInput():
    args = sys.argv
    assert(len(args)==4)
    k,goal,file_name = (args[1],args[2],args[3])
    k = int(k)
    goal = goalEnum(goal)
    file_name = file_name
    return(k,goal,file_name)

#kmeanspp logic:
def build_probabilities (D, p):
    s = sum(D)
    for i in range(len(D)):
        if s!=0:
            p[i] = D[i] /s
    return p

def squared_distance (u, v):
    res = 0
    for i in range(len(u)):
        res += (u[i]-v[i])**2
    return res

def updateDi (datapoints, i, Z, centroids, D):
    tmp = squared_distance(datapoints[i], centroids[Z-1])
    D[i] = min(D[i], tmp)

def first_selection(datapoints, selected_indexes):
    index = np.random.choice(len(datapoints))
    u1 = np.copy(datapoints[index])
    selected_indexes[0] = index
    return u1

def kmeanspp(datapoints, number_of_clusters):
    k = number_of_clusters
    #print("k is "+str(k))
    initial_centroids = [0]*k
    selected_indexes = [0]*k
    initial_centroids[0] = first_selection(datapoints, selected_indexes)
    Z = 1
    D = [squared_distance(datapoints[i],initial_centroids[0]) for i in range(len(datapoints))]
    probabilities = [0] * len(D)
    while Z < k:
        #print("iteration "+str(Z)+"...")
        for i in range(len(datapoints)):
            updateDi (datapoints, i, Z, initial_centroids, D)
        build_probabilities(D, probabilities)
        j = np.random.choice(range(0,len(datapoints)), p = probabilities)
        selected_indexes[Z] = j
        initial_centroids[Z] = np.copy(datapoints[j])
        Z += 1
    return (selected_indexes, [array.tolist() for array in initial_centroids])

#calculate actual centers based on the map: vector_i belongs to cluster_j
def calcCentroidsBasedOnMap (data, data_to_clusters_map, clustersNumber):
    sums = [[0]*len(data[0]) for i in range(clustersNumber)] 
    amounts = [0]*clustersNumber
    for i in range(len(data)):
        for j in range(len(data[0])):
            sums[data_to_clusters_map[i]][j] += data[i][j]
        amounts[data_to_clusters_map[i]] += 1
    for i in range(clustersNumber):
        for j in range(len(data[0])):
            if amounts[i]!=0:
                sums[i][j] /= amounts[i]
    return sums

#main method:
def main():
    np.random.seed(0)
    try:
        k, goal, file_name = getInput()
        pd_data = pd.read_csv(file_name, sep=",", header=None)
        data = pd.read_csv(file_name, sep=",", header=None).to_numpy().tolist()
        n = len (data)
        d = len(data[0])
        if goal == goalEnum.wam:
            result = sp.wam(data, n, d, k)
        elif goal == goalEnum.ddg:
            result = sp.ddg(data, n, d, k)
        elif goal == goalEnum.lnorm:
            result = sp.lnorm(data, n, d, k)
        elif goal == goalEnum.jacobi:
            result = sp.jacobi(data, n, d, k)
        elif goal == goalEnum.spk:
            Rnk,Rnk_to_Data = sp.spk(data, n, d, k)
            numpyData = np.array(Rnk)
            clustersNumber = len(Rnk[0])
            #print("strting choose initials...")
            initial_indexes,initial_vectors = kmeanspp(numpyData,clustersNumber)
            #print("going back to c....")
            vectors_to_clusters_map = sp.kmeans(Rnk , initial_vectors, 1000)
            result = calcCentroidsBasedOnMap(data, vectors_to_clusters_map, clustersNumber) 
            result = {'initial_centroids_indexes':initial_indexes, 'final_centroids':result}
    except ValueError:
        print("Invalid Input!")
        return
    #except:
    #    print("An Error Has Occured")
    #    return
    else:
        if goal ==goalEnum.spk:
            a = [str(y) for y in result['initial_centroids_indexes']]
            b = result['final_centroids']
            print(','.join(a))
            for x in b:
                x0 = ['%.4f'%y for y in x]
                print(','.join(x0))
        else:
            for x in result:
                x0 = ['%.4f'%y for y in x]
                print(','.join(x0))



#RUN:
main()

File: spkmeans.c
#include "spkmeans.h"

#include "res/tools.c"
#include "res/wam.c"
#include "res/ddg.c"
#include "res/lnorm.c"
#include "res/jacobi.c"
#include "res/spk.c"
#include "res/eigenpap.c"
#include "res/fit.c"
#include "res/c_interface.c"

void printArr(double ** A, int n, int m) {
	int i, j;
	for (i = 0; i < n; ++i) {
		for (j = 0; j < m; ++j) {
			printf("%.4f", A[i][j]);
			j + 1 == m ? printf("\n") : printf(",");
		}
	}
}

int main(int argc, char *argv[]) {
	int i, k, data_length, features, *original_to_Rnk_map, error;
	enum goal e;
	double *eigenvalues;
	double **eigenvectors, **initial_centroids, **final_centroids, 
		**data, **_wam, **_ddg, **_lnorm, **_Rnk;
	double  ***_jacobi;
	value_vector_map *vvmap;

	/*Set pointers to NULL, so free() method could
	 *ignore if a pointer was or was not used
	 */
	eigenvalues = NULL;
	eigenvectors = initial_centroids = final_centroids =
		data = _wam = _ddg = _lnorm = _Rnk = NULL;
	_jacobi = NULL;
	original_to_Rnk_map = NULL;

	if (!input(argc, argv, &k, &e, &data, &data_length, &features)) {
		printf("Invalid Input!\n");
		return 0;
	}
	error = 0;
	switch (e)
	{
	case wam:
		_wam = calcWAM(data, data_length, features); 
		if (_wam == NULL) { error = 1; break;}
		printArr(_wam, data_length, data_length);
		break;

	case ddg:
		_wam = calcWAM(data, data_length, features);
		if (_wam == NULL) { error = 1; break; }
		_ddg = calcDDG(_wam, data_length);
		if (_ddg == NULL) { error = 1; break; }
		printArr(_ddg, data_length, data_length);
		break;
	case lnorm:
		_wam = calcWAM(data, data_length, features);
		if (_wam == NULL) { error = 1; break; }
		_ddg = calcDDG(_wam, data_length);
		if (_ddg == NULL) { error = 1; break; }
		_lnorm = calcNGL(_wam, _ddg, data_length);
		if (_lnorm == NULL) { error = 1; break; }
		printArr(_lnorm, data_length, data_length);

		break;
	case jacobi:
		/*Here data is of Mat(n,n) as an assumption*/
		_jacobi = calcJacobi(data, data_length);
		if (_jacobi == NULL) { error = 1; break; }
		eigenvalues = _jacobi[0][0];
		eigenvectors = _jacobi[1];
		printArr(eigenvectors, data_length, data_length);
		printArr(&eigenvalues, 1, data_length);
		break;
	case spk:
		_wam = calcWAM(data, data_length, features);
		if (_wam == NULL) { error = 1; break; }
		_ddg = calcDDG(_wam, data_length);
		if (_ddg == NULL) { error = 1; break; }
		_lnorm = calcNGL(_wam, _ddg, data_length);
		if (_lnorm == NULL) { error = 1; break; }
		_jacobi = calcJacobi(_lnorm, data_length);
		if (_jacobi == NULL) { error = 1; break; }
		eigenvalues = _jacobi[0][0];
		eigenvectors = _jacobi[1];
		vvmap = setMap(eigenvectors, eigenvalues, data_length);
		if (vvmap == NULL) { error = 1; break; }
		if (k == 0) k = determineK(vvmap, data_length);
		_Rnk = calcNormalaizedRnk(vvmap, data_length, k);
		if (_Rnk == NULL) { error = 1; break; }
		initial_centroids = extractCentroids(_Rnk, k);
		if (initial_centroids == NULL) { error = 1; break; }
		original_to_Rnk_map = k_mean(_Rnk, initial_centroids, data_length, k, k, 300);
		if (original_to_Rnk_map == NULL) { error = 1; break; }
		final_centroids = calcFinalCentroids(data, original_to_Rnk_map, k, data_length, features);
		if (final_centroids == NULL) { error = 1; break; }
		printArr(final_centroids, k, features);
		break;
	default:
		error = 1;
		break;
	}
	
	if (error == 1) { printf("An Error Has Occured\n"); return 0; }

	/*free data*/
	if (data != NULL)for (i = 0; i < data_length; ++i)free(data[i]);
	free(data);
	/*free wam*/
	if (_wam != NULL) for (i = 0; i < data_length; ++i)free(_wam[i]);
	free(_wam);
	/*free ddg*/
	if (_ddg != NULL) for (i = 0; i < data_length; ++i)free(_ddg[i]);
	free(_ddg);
	/*free lnorm*/
	if (_lnorm != NULL) for (i = 0; i < data_length; ++i)free(_lnorm[i]);
	free(_lnorm);
	/*free jacobi*/
	if (_jacobi != NULL) {
		free(_jacobi[0][0]); free(_jacobi[0]);
		for (i = 0; i < data_length; ++i)free(_jacobi[1][i]);
	}
	free(_jacobi);
	/*free vvmap*/
	free(vvmap);
	/*free Rnk*/
	if (_Rnk != NULL)for (i = 0; i < data_length; ++i)free(_Rnk[i]);
	free(_Rnk);
	/*free initial_centroids*/
	if (initial_centroids != NULL) for (i = 0; i < k; ++i) free(initial_centroids[i]);
	free(initial_centroids);
	/*free original_to_Rnk_map*/
	free(original_to_Rnk_map);
	/*free final_centroids*/
	if (final_centroids != NULL) for (i = 0; i < k; ++i) free(final_centroids[i]);
	free(final_centroids);
	return 0;
}

double **extractCentroids(double **data, int k) {
	int i, j;
	double **centroids;
	centroids = (double**)malloc(k * sizeof(double*));
	if (centroids == NULL) return NULL;
	for (i = 0; i < k; ++i) {
		centroids[i] = (double*)malloc(k * sizeof(double));
		if (centroids[i] == NULL)return NULL;
		for (j = 0; j < k; ++j) centroids[i][j] = data[i][j];
	}

	return centroids;
}

double ** calcFinalCentroids(double **data, int *map, int k, int data_length, int data_dim) {
	double ** sums;
	int i, j, *amounts;

	sums = (double**)malloc(k * sizeof(double*));
	amounts = (int*)calloc(k, sizeof(int));
	if (sums == NULL || amounts == NULL) return NULL;
	if (sums == NULL) return NULL;
	for (i = 0; i < k; ++i) {
		sums[i] = (double*)calloc(data_dim, sizeof(double));
		if (sums[i] == NULL)return NULL;
	}
	for (i = 0; i < data_length; ++i) {
		for (j = 0; j < data_dim; ++j) {
			sums[map[i]][j] += data[i][j];
			/*sums[map[i]][j] += data[i][j];*/
		}
		amounts[map[i]] += 1;
	}
	for (i = 0; i < k; ++i) {
		for (j = 0; j < data_dim; ++j) {
			if (amounts[i] > 0)sums[i][j] /= amounts[i];
		}
	}

	free(amounts);
	return sums;
}